src/main/java/bio/terra/service/resourcemanagement/ResourceService.java

```
@@ -103,6 +118,86 @@ public GoogleBucketResource getOrCreateBucketForFile(Dataset dataset,
+     */
+    public AzureStorageAccountResource getOrCreateStorageAccount(Dataset dataset,
+                                                                 BillingProfileModel billingProfile,
+                                                                 String flightId) throws InterruptedException {
+        final AzureRegion region =
+            (AzureRegion) dataset.getDatasetSummary().getStorageResourceRegion(AzureCloudResource.STORAGE_ACCOUNT);
+        // Every storage account needs to live in a deployed application's managed resource group, so we make sure that
+        // the application deployment is registered first
+        final AzureApplicationDeploymentResource applicationResource =
+            applicationDeploymentService.getOrRegisterApplicationDeployment(billingProfile);
+
+        List<UUID> storageAccountsForDataset =
+            datasetStorageAccountDao.getStorageAccountResourceIdForDatasetId(dataset.getId());
+
+        String storageAccountName = azureDataLocationSelector.storageAccountNameForDataset(
+            applicationResource.getStorageAccountPrefix(),
+            dataset.getName(),
+            billingProfile);
+
+        List<AzureStorageAccountResource> storageAccountsForBillingProfile = storageAccountsForDataset.stream()
+            .map(this::lookupStorageAccount)
+            .filter(a -> storageAccountIsForBillingProfile(a, billingProfile))
+            .collect(Collectors.toList());
+
+        // there should never be more than one storage account per dataset / billing profile,
+        // so we just take the first one
+        if (storageAccountsForBillingProfile.size() > 0) {
+            AzureStorageAccountResource storageAccount = storageAccountsForBillingProfile.get(0);
+            storageAccountName = storageAccount.getName();
+
+            if (storageAccountsForBillingProfile.size() > 1) {
+                logger.warn("Found more than one storage account associated with this dataset and billing profile");
```

If there should never be more than one billing account, should we throw instead of just warning?

---

I actually think there might be legit usecases for this (e.g. if you're ingesting with different billing profiles...I'm not sure if that should be a failure...)

---

🤷